### PR TITLE
Add Data initializer with UnsafePointer

### PR DESCRIPTION
**Pull request by antonmes**
for example: _expired_

### CODE REVIEW

**Comment by Danappelxx** [comment]
My only concern is that this can lead to performance issues since the contents of the buffer are _copied_ upon initialization of `Data` (line 195). Instead, we should make `UnsafeBufferPointer` conform to `BinaryStream` by using it's `makeIterator` function to do this lazily. 

Snippet to prove this (excuse my swift 2, this was done in a playground):

``` swift
// allocate 3 ints
let pointer = UnsafeMutablePointer<Int>.alloc(3)
pointer.advancedBy(0).memory = 5
pointer.advancedBy(1).memory = 6
pointer.advancedBy(2).memory = 7

let buffer = UnsafeMutableBufferPointer(start: pointer, count: 3)

// create array from buffer
let arr = Array(buffer)

// deallocate those 3 ints
pointer.dealloc(3)

// does not crash, since array _copied_ the contents of the buffer
// rather than just pointing to its contents
arr[0] 
```

**Comment by paulofaria** [comment]
@antonmes did you benchmark this?

File: Sources/Data.swift [hunk 1]
```diff
@@ -190,6 +190,13 @@ extension String: DataConvertible {
 }
 
 extension Data {
+    public init(start: UnsafePointer<Byte>, count: Int) {
+        let buffer = UnsafeBufferPointer(start: start, count: count);
```
Comment (Danappelxx) [inline Sources/Data.swift]: A semicolon? 😱

Comment (antonmes) [inline Sources/Data.swift]: 🙈

**Comment by tanner0101** [comment]
Closing this due to inactivity and merge conflicts.

**Comment by czechboy0** [comment]
Would love to see this resurrected, I'm constantly having to do this manually when working with C code.

**Comment by paulofaria** [comment]
@antonmes can you solve the conflicts?

**Comment by tanner0101** [comment]
Okay, @open-swift/contributors give this a thumbs up if you support.

**Comment by paulofaria** [comment]
@tannernelson there other people that should vote that are not on the github organization. I think we should remind people on slack as well.

**Comment by antonmes** [comment]
swift

**Comment by Danappelxx** [comment]
Aren't the performance issues I mentioned still an issue? I'm not super comfortable with merging this since it has a large underlying performance cost.

**Comment by paulofaria** [comment]
what would be the other option?

**Comment by Danappelxx** [comment]
We could make Data use `AnySequence<Byte>` instead of `Array<Byte>`. That way we could store the raw buffer pointer rather than copying it's data into an array first. This would also make other operations such as working with slices much cheaper.

**Comment by czechboy0** [comment]
@Danappelxx I'm actually trying to solve something similar here: https://github.com/Zewo/gzip/issues/1, so avoiding copying (when safe) would be great.

**Comment by Danappelxx** [comment]
Yeah it's kind of necessary if we want to keep performance a priority. I'm trying to think of how this could be done safely, though. Storing an UnsafeBufferPointer without making it optional is potentially a runtime error.

**Comment by paulofaria** [comment]
awesome! @Danappelxx when you have something send a PR and link this one to it.

**Comment by paulofaria** [comment]
Maybe we could have something like `UnsafeData` which is a fast, unsafe, reference value and make it easily convertible to the slow, safe, value type `Data`.

**Comment by paulofaria** [comment]
like make `UnsafeData` be `DataConvertible` and `Data` be `UnsafeDataConvertible`.

**Comment by Danappelxx** [comment]
That sounds like a good idea! I can send a pr in a couple hours with the `AnySequence` and `UnsafeData` if you want.

**Comment by czechboy0** [comment]
I actually would prefer using AnySequence in this Data, as @Danappelxx suggested. I think we can make sure that it's used safely, plus we'll get a big performance win as we won't need to copy large buffers when converting them into Data. Splitting into UnsafeData doesn't actually solve this problem, as most APIs already use Data, so we'd have to convert (thus copy) to Data in most cases anyway.

**Comment by antonmes** [comment]
@Danappelxx yes, it should be copied. The point is to remove boilerplate when working with C code. 
We already copying anything here: `String`, other `[Byte]`, etc, so it's not big deal. If you're sure about your `UnsafeMutablePointer` lifecycle, then use it instead of `Data` and convert to `Data` only when you have to. 

We could change it to `class Data`and use `var bytes: UnsafeMutablePointer` instead of [Byte], 
but i couldn't use `isUniquelyReferencedNonObjC(&self)`, is that possible?

I believe we just should wait for [this guys](https://github.com/apple/swift-evolution/blob/master/proposals/0069-swift-mutability-for-foundation.md#new-value-types).

**Comment by Danappelxx** [comment]
@antonmes, read the last comments me and paulo posted - using `AnySequence` in conjunction with a new reference type'd data is a better solution.

**Comment by antonmes** [comment]
not sure about that

**Comment by Danappelxx** [comment]
I was specifically talking about

> We could change it to class Data and use var bytes: UnsafeMutablePointer instead of [Byte] [...]

Using `AnySequence` is better than `UnsafeMutablePointer`. Also, `UnsafeBufferPointer` is usually better than `UnsafeMutablePointer` since it provides more convenience, etc.

**Comment by antonmes** [comment]
It's not the point. (`UnsafeMutablePointer` or `UnsafeMutableBufferPointer`)
i'm talking about pointer lifecycle.

**Comment by Danappelxx** [comment]
> We already copying anything here: String, other [Byte], etc, so it's not big deal. If you're sure about your UnsafeMutablePointer lifecycle, then use it instead of Data and convert to Data only when you have to.

No, this is the wrong attitude. Copying a string is not expensive because the strings are usually small, optimized, and copy on write only. `Data` wraps around a potentially massive amount of bytes and will be used in performance-heavy situations, passed around many times. You saw yourself how much the copying affects the performance over at Zewo. Simply saying "use `UnsafeMutablePointer`" is wrong too because `Data` is a standard and therefore is shared between many libraries. Hence, at every cross-module function invocation, we are 100% certainly performing a copy (very expensive).

> I'm talking about pointer lifecycle.

I'm not (and I think @paulofaria and @czechboy0 as well). The issues I raised are strictly performance-related and I think we should resolve those.

**Comment by antonmes** [comment]
Exactly, and this is another issue. But copying from the pointer is not differs from current [default initializer](https://github.com/open-swift/C7/blob/4fd707e2c4c6c9985254dc6d0b83b14784b00aeb/Sources/Data.swift#L5).

**Comment by Danappelxx** [comment]
No, the default initializer copies the struct which is copy on write only, so initializing data that way is almost free. However, creating a new array from a pointer is _not_ free and it copies all of the pointer's contents.

**Comment by antonmes** [comment]
The only safe way to use C buffer and not worry about its lifetime is copying.
Moreover, sometimes you just can't control its lifetime when it was allocated on C side and will be automagicaly deallocated at some point later. @Danappelxx  It's not wrong attitude, it's real case.

Currently, if you want to pass `UnsafeBufferPointer` to any `OpenSwift` / `Zewo` api you should write:

``` swift
let buffer = UnsafeBufferPointer<UInt8>(start: UnsafePointer(data), count: length)
let data = Data(Array(buffer))
```

or [invent your own extension](https://github.com/Zewo/HTTPParser/blob/cc59f1dfea12e01baeb2d8725c97485049f17ff3/Sources/Parser.swift#L79) in **each** library which works with С. 

As you see you can initialize Array with `UnsafeBufferPointer`,
ask your self: why should i use an Array wrapper if it's even more restrictive then stdlib type? 
Because the `Data` is used throughout the API, not just for handling heavy data. 
Btw, the reason we used `struct` instead of `typealias Data = [UInt8]` is because we can't extend `typealias` and not because we wanted to hide "non-copy-on-write" Array functionality. 

anyway, i don't think we need to implement `UnsafeData` ourselves:

> I believe we just should wait for [this guys](https://github.com/apple/swift-evolution/blob/master/proposals/0069-swift-mutability-for-foundation.md#new-value-types).

PS: just to be clear, i'm not saying we don't need `UnsafeData` behaviour, i'm saying that it's another issue, this one - when you **need** copying, `UnsafeData` - when you **sure** about pointer **lifetime**.

**Comment by antonmes** [comment]
The real issue against this PR: this "copy" behaviour may differ from [swift value types API](https://github.com/apple/swift-evolution/blob/master/proposals/0069-swift-mutability-for-foundation.md#developer-experience--api).

I'm closing this PR to avoid undefined behaviour in the future, feel free to reopen if you find it useful.

**Comment by czechboy0** [comment]
Hmm, I'm a bit confused here. Isn't unsafe buffer pointer properly managed by ARC, just like UnsafePointer? If so, migrating Data to use it for storage would both avoid copying (which is a huge problem, I'm trying to implement Gzip middleware and the copying is absolutely killing the performance, it's 40times slower to use Data than NSData with it) and could be properly managed. We can even implement our own copy on write, as is described in the Swift Value Types WWDC video. 

If we want Data to be taken seriously by framework creators, I think we need to solve the problem of copying, and this could be one of them (I'd be happy to hear more solutions, but I think we all need to acknowledge it's a big problem).

**Comment by czechboy0** [comment]
Another option would be to depricate C7.Data and just use Foundation.Data once 0069 is implemented. Tbh that's probably my preferred option 😄

**Pull request closed without merging**